### PR TITLE
Add ssh public keys on user-profile page

### PR DESCRIPTION
**Pull request by Lijiaoa**
TODO: 
- [x] check input format together

image

image

image

improve SSH layout:

image

### CODE REVIEW

File: src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx:1]: Please use this header:

```javascript
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

```

Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx:1]: Will modifiy it. And looks like that still have many files write that.

Comment (hzy46) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx:1]: Yes, they are the old headers. We haven't modified them. This one is the new header.

Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx:1]: ok

File: src/webportal/src/app/job-submission/components/tools/job-ssh.jsx [hunk 2]
```diff
@@ -128,36 +117,32 @@ export const JobSSH = ({ extras, onExtrasChange }) => {
...
-              onSshKeysChange={_onSshKeysGenerated}
+        <Stack gap='l1'>
+          <Hint>
+            Your pre-defined SSH public keys on the{' '}
+            <a href='/user-profile.html'>User Profile</a> page will be set
+            automatically.
+          </Hint>
+          <Stack horizontal gap='l1'>
+            <TextField
+              Lable='Add additional ssh public key'
```
Comment (hzy46) [inline src/webportal/src/app/job-submission/components/tools/job-ssh.jsx:128]: Please change `ssh` to `SSH`.

File: src/webportal/src/app/job-submission/components/tools/job-ssh.jsx [hunk 3]
```diff
@@ -128,36 +117,32 @@ export const JobSSH = ({ extras, onExtrasChange }) => {
...
+              placeholder='Additional ssh public key'
```
Comment (hzy46) [inline src/webportal/src/app/job-submission/components/tools/job-ssh.jsx:129]: Please change `ssh` to `SSH`.

File: src/webportal/src/app/user/fabric/user-profile.jsx [hunk 4]
```diff
@@ -210,6 +250,29 @@ const UserProfile = () => {
           >
             <TokenList tokens={tokens} onRevoke={onRevokeToken} />
           </UserProfileCard>
+          <UserProfileCard
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile.jsx:253]: Please move this `UserProfileCard` before the `UserProfileCard` of Tokens.

SSH is supposed to be used more frequently than tokens.

File: src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        dialogContentProps={{
+          type: DialogType.normal,
+          title: 'Delete SSH Public Keys',
+        }}
+        modalProps={{
+          isBlocking: true,
+        }}
+        minWidth={400}
+      >
+        <div>Are you sure you want to delete this ssh key?</div>
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx:130]: Please change `ssh` to `SSH`.

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  const [inputValueError, setInputValueError] = useState('');
+  const [processing, setProcessing] = useState(false);
+  const [title, setTitle] = useState('');
+  const [value, setValue] = useState('');
+
+  const onAddAsync = async () => {
+    if (title.trim() === '') {
+      setInputTitleError('Please input title');
+    } else if (value.trim() === '') {
+      setInputValueError('Please input ssk value');
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:30]: Please change `ssk` to `SSH`.

File: src/webportal/src/app/job-submission/components/tools/job-ssh.jsx (hunk 2 above)
Comment (Lijiaoa) [reply src/webportal/src/app/job-submission/components/tools/job-ssh.jsx:128]: ok

File: src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx (hunk 5 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list.jsx:130]: ok

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx (hunk 6 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:30]: ok

File: src/webportal/src/app/user/fabric/user-profile.jsx (hunk 4 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile.jsx:253]: ok

File: src/webportal/src/app/job-submission/components/tools/job-ssh.jsx (hunk 3 above)
Comment (Lijiaoa) [reply src/webportal/src/app/job-submission/components/tools/job-ssh.jsx:129]: ok

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    } else {
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:31]: Please validate value matches SSH key's format.

You can use this regex: `/^ssh-rsa AAAA[0-9A-Za-z+/]+[=]{0,3}.*$/`

File: src/webportal/src/app/user/fabric/user-profile.jsx [hunk 8]
```diff
@@ -139,6 +146,39 @@ const UserProfile = () => {
     });
   });
 
+  // click `add public ssh keys button` -> open dialog
+  const onAddPublicKeys = useCallback(async sshPublicKeys => {
+    setSSHProcessing(true);
+    let updatedSSHPublickeys = [];
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile.jsx:152]: Please validate here: the user cannot add ssh keys with duplicate titles. (Title should be unique for one user)

File: src/webportal/src/app/user/fabric/user-profile.jsx [hunk 9]
```diff
@@ -139,6 +146,39 @@ const UserProfile = () => {
...
+  });
+
+  const onDeleteSSHkeys = useCallback(async sshPublicKeys => {
+    let updatedSSHPublickeys = [];
+    if (userInfo.extension.sshKeys) {
+      updatedSSHPublickeys = cloneDeep(userInfo.extension.sshKeys);
+    }
+    updatedSSHPublickeys = updatedSSHPublickeys.filter(
+      item =>
+        item.title !== sshPublicKeys.title &&
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile.jsx:174]: Please only filter with `item.title !== sshPublicKeys.title`.

Title should be unique and delete means delete an ssh key with a certain title.

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        }}
+        modalProps={{
+          isBlocking: true,
+        }}
+        minWidth={600}
+      >
+        <div>
+          <div className={t.mt1}>
+            <TextField
+              label='title'
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:65]: Better to give the user some hints.

`title` -> `Title (give the SSH key a name): `

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+              onChange={e => {
+                setTitle(e.target.value);
+                setInputTitleError(null);
+              }}
+              validateOnFocusOut={true}
+            />
+          </div>
+          <div className={t.mt1}>
+            <TextField
+              label='value'
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:77]: Better to give the user some hints.

Value (SSH Public key, starts with `ssh-rsa`):

File: src/webportal/src/app/user/fabric/conn.js [hunk 12]
```diff
@@ -179,6 +179,26 @@ export const getUserRequest = async username => {
...
+  return fetchWrapper(url, {
+    method: 'PUT',
+    headers: {
+      Authorization: `Bearer ${token}`,
+    },
+    body: JSON.stringify({
+      data: {
+        username: username,
+        extension: {
+          sshKeys: sskMessage,
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/conn.js:194]: `sskMessage` ->  `sshMessage`?

File: src/webportal/src/app/user/fabric/conn.js [hunk 13]
```diff
@@ -179,6 +179,26 @@ export const getUserRequest = async username => {
   });
 };
 
+export const updateUserRequest = async (username, sskMessage) => {
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/conn.js:182]: `sskMessage` ->  `sshMessage`?

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx (hunk 10 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:65]: fixed

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx (hunk 11 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:77]: fixed

File: src/webportal/src/app/user/fabric/conn.js (hunk 12 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/conn.js:194]: updated

File: src/webportal/src/app/user/fabric/conn.js (hunk 13 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/conn.js:182]: updated

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx (hunk 7 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:31]: done

File: src/webportal/src/app/user/fabric/user-profile.jsx (hunk 9 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile.jsx:174]: ok

File: src/webportal/src/app/user/fabric/user-profile.jsx (hunk 8 above)
Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile.jsx:152]: done

File: src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        });
+      } catch (error) {
+        setError(error.message);
+      } finally {
+        setProcessing(false);
+        onDismiss();
+      }
+    }
+
+    // if (title.trim() === '') {
```
Comment (hzy46) [inline src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:66]: Please remove these comments

Comment (Lijiaoa) [reply src/webportal/src/app/user/fabric/user-profile/ssh-list-dialog.jsx:66]: ok

**Review by hzy46 — APPROVED** [review]

**Pull request merged**